galley/pkg/server/process/host.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package process
+
+import (
+	"errors"
+	"sync"
+)
+
+// Host is a process host that takes care of start/stop of sub-components.
+type Host struct {
+	mu         sync.Mutex
+	started    bool
+	components []Component
+}
+
+// Add a new component to this host. It panics if the host is already started.
+func (h *Host) Add(c Component) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if h.started {
+		panic("Host.Add: host is already started")
```

Comment:
Why panic instead of returning error?

Why panic in Add() but return an error in Start()?

Reply:
Start is a runtime concern, whereas Add is a programming / build time, form usage-pattern perspective.
As an example, we can start/stop components numerous times (which happens in the test frameworks), but we would be add'ing things only once. The original model I had for this was to create the Host in one shot through a variadic function that takes components, but it looked uglier, usage-wise, when there are optional components.